From a04e6933183c103f338363dcd55b1fd1132af77a Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Sat, 11 May 2019 19:47:21 +0200
Subject: [PATCH 113/345] MIPS: OCTEON: Add utility helper function
 octeon_read_ptp_csr()

... used by following Ethernet PTP patches.

Signed-off-by: David Daney <david.daney@cavium.com>
---
 arch/mips/include/asm/octeon/octeon.h | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/arch/mips/include/asm/octeon/octeon.h b/arch/mips/include/asm/octeon/octeon.h
index 60481502826a..f2756ce669a6 100644
--- a/arch/mips/include/asm/octeon/octeon.h
+++ b/arch/mips/include/asm/octeon/octeon.h
@@ -8,6 +8,7 @@
 #ifndef __ASM_OCTEON_OCTEON_H
 #define __ASM_OCTEON_OCTEON_H
 
+#include <linux/irqflags.h>
 #include <asm/octeon/cvmx.h>
 #include <asm/bitfield.h>
 
@@ -352,6 +353,25 @@ extern struct cvmx_bootinfo *octeon_bootinfo;
 
 extern uint64_t octeon_bootloader_entry_addr;
 
+static inline uint64_t octeon_read_ptp_csr(u64 csr)
+{
+	if (OCTEON_IS_MODEL(OCTEON_CN63XX_PASS1_X)) {
+		u64 result;
+		unsigned long flags;
+		/*
+		 * CN63XX pass 1.x has an errata where you must read
+		 * this register twice to get the correct result.
+		 */
+		local_irq_save(flags);
+		cvmx_read_csr(csr);
+		result = cvmx_read_csr(csr);
+		local_irq_restore(flags);
+		return result;
+	} else {
+		return cvmx_read_csr(csr);
+	}
+}
+
 extern void (*octeon_irq_setup_secondary)(void);
 
 typedef void (*octeon_irq_ip4_handler_t)(void);
-- 
2.25.1

